proposals/20201026-creating-control-plane.md

```diff
@@ -0,0 +1,433 @@
+}
+
+// +k8s:deepcopy-gen=true
+type PatchSpec struct {
+	Patches []*runtime.RawExtension `json:"patches,omitempty"`
+}
+```
+
+If the designated component's providers do not allow patching, the `PatchSpec` will be ignored. All Nested CRs need to contain a `NestedComponentSpec`. We will not discuss how NCPT and NCP controller generates the `NestedComponentSpec` in this proposal. However, a possible way is to add three fields to the NCPT CRD, and has each of them holds the `NestedComponentSpec` for one of the components. The NCPT will look like the following
+
+```go
+// NestedControlPlaneTemplateSpec defines the desired state of NestedControlPlaneTemplate
+type NestedControlPlaneTemplateSpec struct {
+    // other fields ...    
+    
+    // NestedEtcd contains information required by the etcd provider to create the Etcd
+    NestedEtcd NestedComponentSpec `json:"nestedEtcd"`
+
+    // NestedKAS contains information required by the apiserver provider to create the apiserver
+    NestedKAS NestedComponentSpec `json:"nestedKAS"`
+
+    // NestedKCM contains information required by the controller-manager provider to create the controller-manager
+    NestedKCM NestedComponentSpec `json:"nestedKCM"`
+}
+```
+
+We will let users specify component providers through NCP, implement a native mode that has each sub-controller to create the component on the super cluster, and define three new types `EtcdProvider`, `KASProvider`, and the `KCMProvider`:
+
+```go
+type KASProvider string
+
+type EtcdProvider string
+
+type KCMProvider string
+
+const (
+    NativeKASProvider KASProvider = "NativeKASProvider"
+
+    NativeEtcdProvider  EtcdProvider = "NativeEtcdProvider"
+    EtcdClusterOperator EtcdProvider = "EtcdClusterOperator"
+    EtcdOperator        EtcdProvider = "EtcdOperator"
+
+    NativeKCMProvider KCMProvider = "NativeKCMProvider"
+)
+```
+
+We will add three new fields `NestedEtcdProvider`, `NestedKASProvider`, and the `NestedKCMProvider` into the NCP, which will make the NCP look like the following:
+
+```go
+// NestedControlPlaneSpec defines the desired state of NestedControlPlane
+type NestedControlPlaneSpec struct {
+    // other fields.... 
+    
+    // NestedEtcdProvider specifies which EtcdProvider will be used to create the Etcd
+    // +optional
+    NestedEtcdProvider EtcdProvider `json:"etcdProvider,omitempty"`
+     
+    // NestedKASProvider specifies which KASProvider will be used to create the kube-apiserver
+    // +optional
+    NestedKASProvider KASProvider `json:"kasProvider,omitempty"`
+ 
+    // NestedKCMProvider specifies which KCMProvider will be used to create the kube-controller-manager
+    // +optional
+    NestedKCMProvider KCMProvider `json:"kcmProvider,omitempty"`
+}
+```
+
+### Bootstrap
+
+#### Create prerequisites 
+
+The NCP controller will need to create some resources before sub-controllers start creating the real NCP components.
+
+* **ClusterNamespace** is used to store all resources for the NCP.
```

Comment:
From a Namespacing perspective what do you think about Namespaces being done manually, meaning if you are deploying an NCP into a cluster you create the namespace and create the CRs in that namespace meaning we don't need to deal with Namespace auth delegation from each one of the controllers. Then when we come back around to templating we can add back in the automated Namespace creation into the NCPT, would that work? What do you think?

Reply:
I think in this proposal, we don't need to rely on NCP controller to create all CRs and namespace. We can just assume they are there when controller work. We can leave the discussion to NCP design.

Reply:
Agree, that makes the most sense to me. We'll need to update line 187 to reflect that. 💯